signer-service/src/api/services/sep10.service.js

```diff
@@ -0,0 +1,95 @@
+const { CLIENT_SECRET } = require('../../constants/constants');
+
+const NETWORK_PASSPHRASE = Networks.PUBLIC;
+
+exports.signSep10Challenge = async (challengeXDR, outToken, clientPublicKey) => {
+  const clientDomainStellarKeypair = Keypair.fromSecret(CLIENT_SECRET);
+
+  const { signingKey: anchorSigningKey } = await fetchTomlValues(TOKEN_CONFIG[outToken].tomlFileUrl);
+
+  const transactionSigned = new TransactionBuilder.fromXDR(challengeXDR, NETWORK_PASSPHRASE);
+  if (transactionSigned.source !== anchorSigningKey) {
+    throw new Error(`Invalid source account: ${transactionSigned.source}`);
+  }
+  if (transactionSigned.sequence !== '0') {
+    throw new Error(`Invalid sequence number: ${transactionSigned.sequence}`);
+  }
+
+  // See https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md#success
+  // memo field should be empty as we assume (in this implementation) that we use the ephemeral
+  // to authenticate. But no memo sub account derivation.
+  if (transactionSigned.memo.value === '') {
+    throw new Error('Memo does not match');
+  }
+
+  const { operations } = transactionSigned;
+  // Verify the first manage_data operation
+  const firstOp = operations[0];
+  if (firstOp.type !== 'manageData') {
+    throw new Error('The first operation should be manageData');
+  }
+
+  // Only authorize a session that corresponds with the ephemeral client account
+  if (firstOp.source !== clientPublicKey) {
+    throw new Error('First manageData operation must have the client account as the source');
+  }
+
+  const { anchorExpectedKey: expectedKey, clientDomainEnabled } = TOKEN_CONFIG[outToken];
+  if (firstOp.name !== expectedKey) {
+    throw new Error(`First manageData operation should have key '${expectedKey}'`);
+  }
+  if (!firstOp.value || firstOp.value.length !== 64) {
+    throw new Error('First manageData operation should have a 64-byte random nonce as value');
+  }
+
+  // Flags to check presence of required operations
+  let hasWebAuthDomain = false;
+  let hasClientDomain = false;
+
+  // Verify extra manage_data operations, web_auth and proper client domain.
+  for (let i = 1; i < operations.length; i++) {
+    const op = operations[i];
+
+    if (op.type !== 'manageData') {
+      throw new Error('All operations should be manage_data operations');
+    }
+
+    // Verify web_auth_domain operation
+    if (op.name === 'web_auth_domain') {
+      hasWebAuthDomain = true;
+      if (op.source !== anchorSigningKey) {
+        throw new Error('web_auth_domain manage_data operation must have the server account as the source');
+      }
+    }
+
+    if (op.name === 'client_domain') {
+      hasClientDomain = true;
+      if (op.source !== clientDomainStellarKeypair.publicKey()) {
+        throw new Error('client_domain manage_data operation must have the client domain account as the source');
+      }
+    }
+  }
+
+  //  the web_auth_domain and client_domain operation must be present
+  if (!hasWebAuthDomain) {
```

Comment:
According to [this section](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md#abstract), it does not seem to be mandatory that the challenge transaction has a `web_auth_domain` entry.

Reply:
I understand from [here](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md#success) that is not optional. You get that idea from step 7?

Reply:
It sounds contradictory, I saw that. I thought that this is some backward compatible definition, i.e., an older version did not have this. Whoever creates the challenge should add this field in the current version. Whoever checks the challenge should accept also the older challenge without that field.

But who knows. We can just keep it as it is as long as it works with both anchors.